src/lib/coda_base/sparse_ledger.ml

```diff
@@ -122,49 +130,157 @@ let sub_account_creation_fee action (amount : Currency.Amount.t) =
 
-let apply_user_command_exn t ({sender; payload; signature= _} : User_command.t)
-    =
-  let sender_idx = find_index_exn t (Public_key.compress sender) in
-  let sender_account = get_exn t sender_idx in
-  let nonce = User_command.Payload.nonce payload in
-  let fee = User_command.Payload.fee payload in
-  assert (Account.Nonce.equal sender_account.nonce nonce) ;
+let apply_user_command_exn t
+    ({signer; payload; signature= _} as user_command : User_command.t) =
   let open Currency in
-  let sender_account =
-    { sender_account with
-      nonce= Account.Nonce.succ sender_account.nonce
-    ; receipt_chain_hash=
-        Receipt.Chain_hash.cons payload sender_account.receipt_chain_hash
-    ; balance=
-        Balance.sub_amount sender_account.balance (Amount.of_fee fee)
-        |> Option.value_exn ?here:None ?error:None ?message:None }
+  let signer_pk = Public_key.compress signer in
+  (* TODO: Put actual value here. See issue #4036. *)
+  let current_global_slot = Coda_numbers.Global_slot.zero in
+  (* Fee-payer information *)
+  let fee_token = User_command.fee_token user_command in
+  let fee_payer = User_command.fee_payer user_command in
+  let nonce = User_command.nonce user_command in
+  let () =
+    (* TODO: Enable multi-sig. *)
+    assert (
+      Public_key.Compressed.equal (Account_id.public_key fee_payer) signer_pk
+    )
   in
-  match User_command.Payload.body payload with
-  | Stake_delegation (Set_delegate {new_delegate}) ->
-      set_exn t sender_idx {sender_account with delegate= new_delegate}
-  | Payment {amount; receiver} ->
-      let t =
-        set_exn t sender_idx
-          { sender_account with
+  let () = assert (Token_id.equal fee_token Token_id.default) in
+  let fee_payer_idx, fee_payer_account =
+    let idx = find_index_exn t fee_payer in
+    let account = get_exn t idx in
+    assert (Account.Nonce.equal account.nonce nonce) ;
+    let fee = User_command.fee user_command in
+    let timing =
+      Or_error.ok_exn
+      @@ Transaction_logic.validate_timing ~txn_amount:(Amount.of_fee fee)
+           ~txn_global_slot:current_global_slot ~account
+    in
+    ( idx
+    , { account with
+        nonce= Account.Nonce.succ account.nonce
+      ; balance=
+          Balance.sub_amount account.balance (Amount.of_fee fee)
+          |> Option.value_exn ?here:None ?error:None ?message:None
+      ; receipt_chain_hash=
+          Receipt.Chain_hash.cons payload account.receipt_chain_hash
+      ; timing } )
+  in
+  (* Charge the fee. *)
+  let t = set_exn t fee_payer_idx fee_payer_account in
+  let source = User_command.source user_command in
+  let receiver = User_command.receiver user_command in
+  let exception Reject of exn in
+  (* Raise an exception if any of the invariants for the user command are not
+     satisfied, so that the command will not go through.
+
+     This must re-check the conditions in Transaction_logic, to ensure that the
+     failure cases are consistent.
+  *)
+  match
```

Comment:
This nested match is awfully difficult to read. Can can the value we are matching on be lifted up into a `let` binding before the match instead?

Reply:
Replaced with a function binding and a `try .. with` wrapping the function call, hopefully it's clearer now.